File: ELEC881_Copilot/src/Function Name Masked/Java script (JS)/q29-longest-increasing-path-in-a-matrix.js
var long = function(matrix) {
    if (matrix.length === 0) return 0;
    let m = matrix.length, n = matrix[0].length;
    let dp = new Array(m).fill(null).map(() => new Array(n).fill(0));
    let dirs = [[0, 1], [1, 0], [0, -1], [-1, 0]];
    let dfs = (i, j) => {
        if (dp[i][j] !== 0) return dp[i][j];
        let max = 1;
        for (let dir of dirs) {
            let x = i + dir[0], y = j + dir[1];
            if (x < 0 || x >= m || y < 0 || y >= n || matrix[x][y] <= matrix[i][j]) continue;
            max = Math.max(max, 1 + dfs(x, y));
        }
        dp[i][j] = max;
        return max;
    };
    let res = 1;
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            res = Math.max(res, dfs(i, j));
        }
    }
    return res;
}

File: ELEC881_Copilot/src/Function Name Masked/Python/q29-longest-increasing-path-in-a-matrix.py
class abc(object):
    def long(self, matrix):
        if not matrix:
            return 0
        m, n = len(matrix), len(matrix[0])
        dp = [[0] * n for _ in range(m)]
        res = 0
        for i in range(m):
            for j in range(n):
                res = max(res, self.dfs(matrix, i, j, dp))
        return res
    
    def dfs(self, matrix, i, j, dp):
        if dp[i][j]:
            return dp[i][j]
        m, n = len(matrix), len(matrix[0])
        for x, y in [(i-1, j), (i+1, j), (i, j-1), (i, j+1)]:
            if 0 <= x < m and 0 <= y < n and matrix[x][y] > matrix[i][j]:
                dp[i][j] = max(dp[i][j], self.dfs(matrix, x, y, dp))
        dp[i][j] += 1
        return dp[i][j]